src/Autonomous.cpp

```diff
@@ -412,30 +416,39 @@ void Autonomous::autonomyIter()
       map.addPoints(toTransform(pose), lidar_scan, mapDoesOverlap ? 0.4 : 0);
       mapDoesOverlap = lidar_scan.size() > mapOverlapSampleThreshold;
     }
-    if (plan_cost >= INFINITE_COST || ++time_since_plan % REPLAN_PERIOD == 0) {
-      // TODO planning should happen in a separate thread
-      // because it takes a long time
-      point_t point_t_goal;
-      point_t_goal.topRows(2) = plan_target.topRows(2);
-      point_t_goal(2) = 1.0;
-      point_t_goal = toTransform(pose) * point_t_goal;
-      //plan_t new_plan = getPlan(lidar_scan, point_t_goal, PLANNING_GOAL_REGION_RADIUS);
-      plan_t new_plan = getPlan([&](double x, double y, double radius)->bool {
-        // transform the point to check into map space
-        point_t relPoint = {x, y, 1};
-        point_t p = invTransform * relPoint;
-        return map.hasPointWithin(p, radius);
-      }, point_t_goal, PLANNING_GOAL_REGION_RADIUS);
-      double new_plan_cost = planCostFromIndex(new_plan, 0);
-      // we want a significant improvement to avoid thrash
-      log(LOG_DEBUG, "old cost %f, new cost %f\n", plan_cost, new_plan_cost);
-      if (new_plan_cost < plan_cost * 0.8) {
-        plan_idx = 0;
-        plan_base = pose;
-        plan_cost = new_plan_cost;
-        plan = new_plan;
-        time_since_plan = 0;
-      }
+    if ((plan_cost >= INFINITE_COST || ++time_since_plan % REPLAN_PERIOD == 0)) {
+		// if we don't have a plan pending, we should start computing one.
+		if (!pending_plan.valid()) {
+			point_t point_t_goal;
+			point_t_goal.topRows(2) = plan_target.topRows(2);
+			point_t_goal(2) = 1.0;
+			point_t_goal = toTransform(pose) * point_t_goal;
+			auto collide_func = [&](double x, double y, double radius) -> bool {
+				// transform the point to check into map space
+				point_t relPoint = {x, y, 1};
+				point_t p = invTransform * relPoint;
+				return map.hasPointWithin(p, radius);
+			};
+			pending_plan =
+				std::async(std::launch::async, &computePlan, collide_func, point_t_goal);
+		}
+		// if there is a plan pending, check if it is ready.
+		else if(pending_plan.wait_for(std::chrono::milliseconds(0)) == std::future_status::ready){
```

Comment:
I think this block should be moved outside the parent if statement so that we can check if a plan is ready on every iteration and choose better plans faster. The current behavior is slow to adopt new plans and start replanning.

Reply:
Yeah, I think now that planning is a separate thread we can just have that thread constantly be running. Obviously we should make sure not to launch a bunch of overlapping plan computations but as long as we only have one `pending_plan` member variable I think this should be fine.